# DSCI 235 Final Exam Spring 22

In this exam you will analyze several datasets using the tools we have studied this semester.

**Instructions:** 

* You have two hours 
* Sign and date the honor statement below.
* Do not use loop constructs unless explicitly suggested.
* Submit your notebook via Canvas


Honor statement:

**I affirm that I have not given or received any unauthorized help on this exam, and that this work is my own.**

Matthew Boyd - May 2, 2022

In [1]:
import pandas as pd

## Part 1:  The world happiness report

The World Happiness Report is a survey of the state of global happiness. 
It has been published regularly since 2012, and the focus here is on the results of the third report from 2015.  The 2015 report ranks 158 countries by their happiness levels.
The report continues to gain global recognition as governments, organizations and civil society increasingly use happiness indicators to inform their policy-making decisions.

Each country has the following information:

* Happiness score - a number between 1 and 10 that represents the average polled happiness score for that country.
* Happiness rank:  the rank of each country in terms of its happiness score (the country with the highest happiness score has happiness rank equal to 1).
* Region - the region the country belongs to.
* Other information that is correlated with happiness (life expectancy, GDP, sense of freedom, the importance of family, etc.)

In our analysis we will focus on the happiness score/rank and region information.

Let's read the data:

In [2]:
happiness = pd.read_csv("https://www.cs.colostate.edu/~asa/courses/dsci235/spr21/data/2015.csv")
happiness.head()

Country          Region  Happiness Rank  Happiness Score  \
0  Switzerland  Western Europe               1            7.587   
1      Iceland  Western Europe               2            7.561   
2      Denmark  Western Europe               3            7.527   
3       Norway  Western Europe               4            7.522   
4       Canada   North America               5            7.427   

   Standard Error  Economy (GDP per Capita)   Family  \
0         0.03411                   1.39651  1.34951   
1         0.04884                   1.30232  1.40223   
2         0.03328                   1.32548  1.36058   
3         0.03880                   1.45900  1.33095   
4         0.03553                   1.32629  1.32261   

   Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
0                   0.94143  0.66557                        0.41978   
1                   0.94784  0.62877                        0.14145   
2                   0.87464  0.64938                        0.48357   
3                   0.88521  0.66973                        0.36503   
4                   0.90563  0.63297                        0.32957   

   Generosity  Dystopia Residual  
0     0.29678            2.51738  
1     0.43630            2.70201  
2     0.34139            2.49204  
3     0.34699            2.46531  
4     0.45811            2.45176

Countries belong to the following regions:

In [3]:
happiness['Region'].unique()

array(['Western Europe', 'North America', 'Australia and New Zealand',
       'Middle East and Northern Africa', 'Latin America and Caribbean',
       'Southeastern Asia', 'Central and Eastern Europe', 'Eastern Asia',
       'Sub-Saharan Africa', 'Southern Asia'], dtype=object)

Using this data answer the following:

* Which nation is the happiest among the nations of Latin America and the Caribbean? What is that country's happiness rank?

In [8]:
happiness[(happiness['Region'] == "Latin America and Caribbean")].sort_values('Happiness Score', ascending = False).head(1)

Country                       Region  Happiness Rank  Happiness Score  \
11  Costa Rica  Latin America and Caribbean              12            7.226   

    Standard Error  Economy (GDP per Capita)   Family  \
11         0.04454                   0.95578  1.23788   

    Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
11                   0.86027  0.63376                        0.10583   

    Generosity  Dystopia Residual  
11     0.25497            3.17728

Costa Rica is happiest in Latin America and Caribbean with a hapiness score of 7.2 and rank of 12.

* How many countries have a happiness score above 7?

In [10]:
len(happiness[(happiness['Happiness Score'] > 7)])

15

There are 15 countries with a happiness score above 7.

* Compute the mean happiness score for each region.

In [13]:
happiness.groupby('Region')['Happiness Score'].mean()

Region
Australia and New Zealand          7.285000
Central and Eastern Europe         5.332931
Eastern Asia                       5.626167
Latin America and Caribbean        6.144682
Middle East and Northern Africa    5.406900
North America                      7.273000
Southeastern Asia                  5.317444
Southern Asia                      4.580857
Sub-Saharan Africa                 4.202800
Western Europe                     6.689619
Name: Happiness Score, dtype: float64

* For each region we will define the happiness spread as the maximum happiness score in that region minus the minimum happiness score in that region.  For example, for the region of *Australia and New Zealand* the spread is 0.002 (the happiness score of New Zealand - the happiness score of Australia).  Compute the happiness spread for all regions.

In [21]:
min = happiness.groupby('Region')['Happiness Score'].min()
max = happiness.groupby('Region')['Happiness Score'].max()
max - min

Region
Australia and New Zealand          0.002
Central and Eastern Europe         2.287
Eastern Asia                       1.424
Latin America and Caribbean        2.708
Middle East and Northern Africa    4.272
North America                      0.308
Southeastern Asia                  2.979
Southern Asia                      1.678
Sub-Saharan Africa                 2.638
Western Europe                     2.730
Name: Happiness Score, dtype: float64

## Part 2: Analyzing movie ratings

For the following questions we will use a dataset that provides information on movie ratings.

The data comes in two files:  "movies.dat", which describes the movies, and "ratings.dat" which provides the movie ratings.

The following commands will read the data into pandas DataFrames:

In [22]:
movies=pd.read_csv('https://www.cs.colostate.edu/~asa/courses/dsci235/spr21/data/movies.dat', sep='::', engine='python', header=None, names=['movie_id', 'title', 'genres'])
movies.head()

movie_id                               title                        genres
0         1                    Toy Story (1995)   Animation|Children's|Comedy
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy
2         3             Grumpier Old Men (1995)                Comedy|Romance
3         4            Waiting to Exhale (1995)                  Comedy|Drama
4         5  Father of the Bride Part II (1995)                        Comedy

The movies DataFrame we just read has the following information:

- movie_id: an integer between 1 and 3952
- Titles are identical to titles provided by IMDB (including year of release)
- Genres are pipe-separated (we'll ignore them in this work).

Next we will read the ratings data, which contains the following information:

- user_id:  an integer in the range between 1 and 6040 
- movie_id: an integer between 1 and 3952
- rating:  Ratings are on a 5-star scale (whole-star ratings only)
- time_stamp is represented in seconds as returned by Unix `time` command.


In [25]:
ratings=pd.read_csv('https://www.cs.colostate.edu/~asa/courses/dsci235/spr21/data/ratings_subset.dat', sep='::', engine='python', header=None, names=['user_id', 'movie_id', 'rating', 'time_stamp'])
ratings.head()

user_id  movie_id  rating  time_stamp
0            1      1193       5   978300760
1            1       661       3   978302109
2            1       914       3   978301968
3            1      3408       4   978300275
4            1      2355       5   978824291
...        ...       ...     ...         ...
99995      669       585       4   975631958
99996      669       586       3   975631958
99997      669       587       3   975631938
99998      669       588       4   975631606
99999      669       589       5   975633871

[100000 rows x 4 columns]

Write pandas commands that answer the following questions:

* The ratings DataFrame identifies movies by their id.  To obtain movie ratings by name, `merge` the two DataFrames (movies and ratings).  Perform an `inner` join when merging.  The resulting DataFrame might have fewer rows than the original ratings DataFrame.  Explain why this is the case.

In [26]:
# merge the DataFrames
df = pd.merge(movies, ratings, how="inner", on=['movie_id'])
df

movie_id             title                       genres  user_id  \
0             1  Toy Story (1995)  Animation|Children's|Comedy        1   
1             1  Toy Story (1995)  Animation|Children's|Comedy        6   
2             1  Toy Story (1995)  Animation|Children's|Comedy        8   
3             1  Toy Story (1995)  Animation|Children's|Comedy        9   
4             1  Toy Story (1995)  Animation|Children's|Comedy       10   
...         ...               ...                          ...      ...   
97615      3813  Interiors (1978)                        Drama      230   
97616      3813  Interiors (1978)                        Drama      319   
97617      3813  Interiors (1978)                        Drama      333   
97618      3813  Interiors (1978)                        Drama      550   
97619      3813  Interiors (1978)                        Drama      588   

       rating  time_stamp  
0           5   978824268  
1           4   978237008  
2           4   978233496  
3           5   978225952  
4           5   978226474  
...       ...         ...  
97615       4   976824650  
97616       3  1002940959  
97617       3   996355393  
97618       3   976169951  
97619       3   985648666  

[97620 rows x 6 columns]

The new data frame has less rows because we perfomred an inner join. This join only keeps rows that have a matching value in both data frames being joined. This shows that the 'ratings' data frame has some movies that aren't listed in 'movies'

* Which movie received the largest number of five star ratings?  (In this and other questions, provide titles rather than IDs!)

In [31]:
# largest number of five star ratings
df[df['rating'] == 5].groupby('title').size().sort_values(ascending = False).head(2)

title
American Beauty (1999)                       228
Star Wars: Episode IV - A New Hope (1977)    205
dtype: int64

American Beauty in 1999 received the most 5 star ratings with 228.

* Which movie received the highest average rating?  For this analysis, ignore movies that received fewer than 50 ratings.

In [61]:
# highest average rating
size = df.groupby('title').size()
size = pd.DataFrame(size).reset_index()
large = pd.DataFrame(size[size[0] >= 50])
df1 = pd.merge(df, large, how = 'inner', on=['title'])
df1.groupby('title')['rating'].mean().sort_values(ascending = False).head(2)


title
Seven Samurai (The Magnificent Seven) (Shichinin no samurai) (1954)    4.660714
Schindler's List (1993)                                                4.562249
Name: rating, dtype: float64

Seven Samurai (The Magnificent Seven) (Shichinin no samurai) (1954) had the highest average rating among movies with at least 50 ratings. This movie had a rating of 4.66.

## Part 3:  NumPy

### Counting zeros

For a 1-d array $x$, we'll define its `number_of_zeros` as the number of elements in the array that are equal to zero.
For example, for the array 

```Python
[1, 5, 0, 6, 0, 1]
```

The `number_of_zeros` is equal to 2.

We can apply `number_of_zeros` to a matrix $X$ (i.e. a 2-d array).
The definition can be applied either to the columns or rows of the matrix, resulting in an array of `number_of_zeros` values for each column/row.  Your task is to write a function that computes that.  You may not use the NumPy functions count_nonzero, nonzero, and argwhere.

For example, for the matrix
```Python
2 0 3 0
0 0 1 5
0 0 0 6
```

when applied to the columns, the result should be an array that contains the numbers

```Python
2 3 1 1
```

when applied to the rows the result should be an array that contains the numbers

```Python
2 2 3
```

Fill in the following function for computing `number_of_zeros`.  The axis  argument should control whether the operation is performed on columns or rows.

In [62]:
import numpy as np

In [151]:
def number_of_zeros(X, axis = 0):
  I = np.zeros(np.shape(X))
  return np.sum(I == X, axis = axis)

In [174]:
# test your code here
X = np.array([[3,0], [6,0],[8,2]]) 
print(number_of_zeros(X))
print(number_of_zeros(X, 1))

[0 2]
[1 1 0]


### Removing sparse columns

Write a function that removes sparse columns from a 2-d array.
We will define a sparse column as a column that contains mostly zeros, and more specifically, the number of zeros is at least 90% of the entries in the column.  For example, if we apply this to the matrix

```Python
2 0 3 0
0 0 1 5
0 0 0 6
```

The second column would be removed.
You can use the `number_of_zeros` function you just wrote to help you in this task.

In [205]:
def remove_sparse_columns(X):
  X1 = X == np.zeros(np.shape(X))
  return X[:,np.sum(X1, axis = 0) / len(X1) <= 0.9]
  

In [206]:
X = np.array([[2, 0, 3, 0], [0,0,1,5],[0,0,0,6]]) 
remove_sparse_columns(X)

array([[2, 3, 0],
       [0, 1, 5],
       [0, 0, 6]])